### DISCUSSION TITLE
API Data (Could've been a discussion)

### DISCUSSION THREAD

**Discussion by ArneTR** [dos-group/vessim#224] [Ideas]
Hey folks,

thanks for bringing this project to live. I stumbled over it in the list of HotCarbon and was intrigued by the description as I was looking for a carbon aware testing bed for a while.

Initially I wanted to open a discussion and not an issue, but you do not have that feature activated. Feel free to move it over in case you activate it in the future.

My question: After reading the README and the docs I am not exaclty clear how I would test my application in this test bed.
Assuming I have an application that is already carbon aware, so it will probably use ElectricityMaps, Wattime, CarbonAwareSDK, entso-e or any other current API.
When I put it in the testbed, will it mock the API endpoint automatically and send me the data of the pre-defined to be simulated micro grid? If not, how is the data to be ingested by my application?

**Comment by birnbaum** [comment]
Hi Arne, thanks for your question!

If I understand correctly, you are planning to mock the APIs of the big data providers but instead replay your own data for testing? We don't re-implement any of the big APIs yet, but this can be done via a custom controller. We will soon improve our docs on this, but if you tell me your exact use case we can of course help you!

And thanks for the hint, we activated Github discussions for this repo :)

**Comment by ArneTR** [comment]
My use case is: I have an application that is generally carbon aware, as in it pulls information in about the grid intensity from APIs and then performs demand-shaping or time-shifting .

I was under the assumption that this codebase, since you call it a "testbed for carbon aware applications" could help me test that.

However I have not understood so far what would be the use case that the testbed covers if you do not mock the APIs. 

So maybe I understand the codebase not really atm. Can you maybe explain what the typical use case for it is? Or how my application could use this codebase to test it's carbon aware functionality?

**Comment by birnbaum** [comment]
At this point, Vessim has been mostly designed as a research tool, which is why we haven't re-implemented any of the big APIs yet. The co-simulation concept behind Vessim pays off when you are trying to "mock" an entire (or multiple) microgrids, including producers, energy storage, etc. and want to develop _new_ interfaces to interact with/react to the state of such systems. That means, most use cases go beyond optimizing for carbon intensity (which oversimplifies many real-life effects).

The use cases are also described in Section 2 of our paper: https://arxiv.org/pdf/2306.09774

However, I fully agree that what you are describing is a very basic feature, especially for testing carbon-aware applications in practice. We will soon improve our docs on how to implement REST APIs for interacting with Vessim simulations. Also, we'll consider adding Vessim as a data source for the carbon-aware SDK: https://carbon-aware-sdk.greensoftware.foundation/docs/tutorial-extras/selecting-a-data-source
